fix: #16028 Org Settings - SEOIndexing,Disable Org Url

vijayraghav-io opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #16028 
- Fixes CAL-4097

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

https://www.loom.com/share/5421eda4a55a4c6fbcb690f26bc6e504?sid=bb1ece9f-d52d-480a-b776-cd6898618d0f

https://www.loom.com/share/415c9dab0e54482594e3e7a9994d9772?sid=a313a1cf-db70-4875-8400-6d9457a896b5

https://www.loom.com/share/fb9f0bb399244ecca6f0e6986a2b2fb3?sid=deb7a35d-5bb9-4b62-a98d-39b40994d94d

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] - N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
update orgsettings

vijayraghav-io committed:
update for org seo indexing

Amit91848 commented:
Converting to draft as this looks incomplete.

Amit91848 commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -2519,5 +2519,7 @@
   "outlook_connect_atom_loading_label": "Checking Outlook Calendar",
   "booking_question_response_variables": "Booking question response variables",
+  "disable_org_url_label": "Disable Organization URL and redirect",
+  "disable_org_url_description": "Hide your deel.cal.com subdomain and redirect to deal.com instead.",
```
```suggestion
  "disable_org_url_description": "Hide your deel.cal.com subdomain and redirect to deel.com instead.",
```

vijayraghav-io replied:
will update

CarinaWolli replied:
@Amit91848 shouldn't we use acme.com as an example here?

Amit91848 replied:
I thought so too acme would be better, but the screenshot in issue had deel in description.

Amit91848 commented on packages/features/ee/organizations/pages/settings/appearance.tsx:
```diff
@@ -43,6 +43,12 @@ const OrgAppearanceView = ({
   } = themeForm;
 
   const [hideBrandingValue, setHideBrandingValue] = useState(currentOrg?.hideBranding ?? false);
+  const [allowSEOIndexingValue, setAllowSEOIndexingValue] = useState(
```
Based on the screenshot, we want this on the profile page of organization settings and not appearance. Could also be on the general page of org settings as that is where it is for normal users.

Amit91848 reviewed (commented):
Left some minor NIT

vijayraghav-io commented:
> Converting to draft as this looks incomplete.

@Amit91848 
Except redirecting sub-domain to domain (its in progress i am updating this), rest all are done -
allowSEOIndexing - UI , DB save and retrieval, marking for SEOIndexing based on this org setting (users-type-public-view.getServerSideProps.tsx)
disableOrgSubdomainURL - UI, DB save and retrieval, (redirecting based on this setting is pending)

vijayraghav-io committed:
updated for domain rediredt

vijayraghav-io commented:
updated for subdomain redirect

vijayraghav-io committed:
update for new url

Amit91848 commented on apps/web/lib/getMainDomainOrgRedirect.ts:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  const protocol = WEBAPP_URL.startsWith("https://") ? "https" : "http";
+  const originalUrl = new URL(req.url!, `${protocol}://${req.headers.host}`);
+
+  const domains = originalUrl.hostname.split(".");
+
+  if (domains.length <= 2) {
+    return null;
+  }
+  log.debug(`Looking for redirect from`, originalUrl.toString());
+
+  const newHostName = `${domains[0]}.${domains.at(-1)}`;
+  originalUrl.hostname = newHostName;
```
If url is acme.cal.com, that doesn't necessarily mean their domain has to be acme.com. Can be anything really, imo we should take url input from them and redirect to that url. Similar to how we do it for successful booking redirect.

@PeerRich what do you think about this?

vijayraghav-io replied:
In case, we take input from user, we need to validate against our process.env.ALLOWED_HOSTS.

And in any case whether we take input or just assume `acme.cal.com` to be redirected to `acme.com`, we need to append ALLOWED_HOSTS with '"com","dev","community".

This is because there is some validation happening in orgDomain.ts
https://github.com/vijayraghav-io/cal.com/blob/3a46375fed575dc9a35bea2c534044f3a1f9fb2b/packages/features/ee/organizations/lib/orgDomains.ts#L35

if we just redirect without ALLOWED_HOSTS being appended with "com", this validation will fail, and org slug is not identified correctly.

@PeerRich , can you please share your thoughts on this?
Also, please confirm if we need to accept input from user for redirect.

vijayraghav-io replied:
Since we are redirecting hostname itself, we should not take input from user, this leads to failed validation against ALLOWED_HOSTNAMES and org slug will not identified correctly.

PeerRich replied:
@Amit91848 @vijayraghav-io we need to use the verified "autoInviteURL" for this, @hariombalhara knows

PeerRich replied:
> we should take url input from them and redirect to that url. Similar to how we do it for successful booking redirect.

we should not take url input from them, cause this can lead to an abuse vector (adidas.cal.com => adidas.phishing-website.com) instead we should redirect to their verified Org domain (the one they use for all auto-invites for emails)

it's also one less input

hariombalhara replied:
We have `organizationSettings.orgAutoAcceptEmail` that stores the domain like 'cal.com', 'acme.com'. We should use that itself for the redirect.

Amit91848 replied:
> > we should take url input from them and redirect to that url. Similar to how we do it for successful booking redirect.
> 
> we should not take url input from them, cause this can lead to an abuse vector (adidas.cal.com => adidas.phishing-website.com) instead we should redirect to their verified Org domain (the one they use for all auto-invites for emails)
> 
> it's also one less input


Makes much more sense 👍

Amit91848 reviewed (changes requested):
Thanks for the PR @vijayraghav-io ! I have some doubts about redirecting to org domain.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
updated ALLOWED_HOSTNAMES

vijayraghav-io committed:
updated description

vijayraghav-io committed:
nit/chore: add log

vijayraghav-io committed:
update for user slug

vijayraghav-io committed:
update redirect when team slug is visited

anikdhabal commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -175,6 +176,15 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
     } as const;
   }
 
+  // If parent is org and based on setting, redirect to main domain org.
```
Remove the comment. It is understandable

anikdhabal commented on .env.example:
```diff
@@ -66,7 +66,7 @@ SAML_CLIENT_SECRET_VERIFIER=
 PGSSLMODE=
 
 # Define which hostnames are expected for the app to work on
-ALLOWED_HOSTNAMES='"cal.com","cal.dev","cal-staging.com","cal.community","cal.local:3000","localhost:3000"'
+ALLOWED_HOSTNAMES='"cal.com","cal.dev","cal-staging.com","cal.community","cal.local:3000","localhost:3000","com","dev","community","local:3000"'
```
Why this changes?

vijayraghav-io replied:
this is required to identify the OrgDomain or OrgSlug correctly when `acme.cal.com` is redirected to `acme.com` , acme is recognized as OrgDomain in `acme.com` url

hariombalhara replied:
We shouldn't change this. AFAIU, we don't want to support acme.com. 

acme.com will be the organization's own website and it has nothing to do with us.

vijayraghav-io replied:
if we don't append "com" to ALLOWED_HOSTNAMES,
the validation here fails https://github.com/calcom/cal.com/blob/ad62f9d2bdb56a0374a30c45bd939546de25e8e0/packages/features/ee/organizations/lib/orgDomains.ts#L35
so getOrgSlug() fn will return null, instead of slug

vijayraghav-io replied:
handled this in alternate method , without appending env variable

anikdhabal reviewed (commented):
Need test

anikdhabal commented on apps/web/modules/users/views/users-type-public-view.getServerSideProps.tsx:
```diff
@@ -234,7 +243,8 @@ async function getUserPageProps(context: GetServerSidePropsContext) {
     trpcState: ssr.dehydrate(),
     isBrandingHidden: user?.hideBranding,
-    isSEOIndexable: user?.allowSEOIndexing,
+    isSEOIndexable:
+      user?.profile?.organization?.organizationSettings?.allowSEOIndexing ?? user?.allowSEOIndexing,
```
So, if either of them is true, you allow SEO indexing for both personal and organizational profiles. Is this what we need? Isn't that different?

vijayraghav-io replied:
yeah right this is not required, as org/team slug will take care of org settings, my bad i missed to remove this after updating team slug.

anikdhabal commented:
@vijayraghav-io Also need a Loom video that demonstrates both functionalities working as expected. The provided Loom video only shows the setting changes and database changes.

anikdhabal reviewed (changes requested):
Left some comments

vijayraghav-io committed:
updated users-type-public-view

vijayraghav-io commented:
> @vijayraghav-io Also need a Loom video that demonstrates both functionalities working as expected. The provided Loom video only shows the setting changes and database changes.

@anikdhabal , PFB video for subdomain redirect

https://www.loom.com/share/d7e3a771780d43409c0b64cc0acdee5a?sid=500988fe-efba-4c0d-956a-8a9a52c10201

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
added tests for getMainDomainOrgRedirect

vijayraghav-io committed:
updated tests

vijayraghav-io commented:
@anikdhabal , uploaded videos for both SEOIndexing and Redirect functionality based on settings,
also added tests

SEOIndexing Video
https://www.loom.com/share/fb9f0bb399244ecca6f0e6986a2b2fb3?sid=deb7a35d-5bb9-4b62-a98d-39b40994d94d

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

anikdhabal commented on apps/web/modules/users/views/users-type-public-view.getServerSideProps.tsx:
```diff
@@ -223,6 +232,13 @@ async function getUserPageProps(context: GetServerSidePropsContext) {
     } as const;
   }
 
+  // if Org allows SEOIndexable then only take user's SEOIndexable settings.
+  const allowSEOIndexing = org
+    ? user?.profile?.organization?.organizationSettings?.allowSEOIndexing
+      ? user?.allowSEOIndexing
+      : false
+    : user?.allowSEOIndexing;
+
```
Nah, I don't think this is correct. Currently, the personal profile is also SEO indexable if the organization profile SEO indexing is enabled. Also pls remove comment

anikdhabal commented on apps/web/modules/users/views/users-public-view.tsx:
```diff
@@ -58,6 +58,14 @@ export function UserPage(props: InferGetServerSidePropsType<typeof getServerSide
 
   const isEventListEmpty = eventTypes.length === 0;
   const isOrg = !!user?.profile?.organization;
+
+  // if Org allows SEOIndexable then only take user's SEOIndexable settings.
```
same here

anikdhabal commented on apps/web/lib/getMainDomainOrgRedirect.test.ts:
```diff
@@ -0,0 +1,78 @@
+};
+
+describe("getMainDomainOrgRedirect", () => {
+  // 'users-type-public-view' sub-domain sample with queryparams
+  // "http://orgdomain.cal.com/admin/orgevent?month=2024-08&date=2024-08-09"
```
remove

anikdhabal commented on apps/web/lib/getMainDomainOrgRedirect.test.ts:
```diff
@@ -0,0 +1,78 @@
+    });
+  });
+
+  // 'users-type-public-view' No sub-domain sample with queryparams
+  // "http://orgdomain.com/admin/orgevent?month=2024-08&date=2024-08-09"
```
Remove it and other places also

anikdhabal reviewed (changes requested):
Left some comments

vijayraghav-io committed:
nit/chore

anikdhabal reviewed (dismissed):
LGTM !! Great work @vijayraghav-io 

Need one more approval from @calcom/foundation team before merge

hariombalhara reviewed (changes requested):
There is a requirement to change the redirect. Blocking merge for that

hariombalhara commented on apps/web/lib/getMainDomainOrgRedirect.ts:
```diff
@@ -0,0 +1,33 @@
+import logger from "@calcom/lib/logger";
+
+const log = logger.getSubLogger({ prefix: ["lib", "getMainDomainOrgRedirect"] });
+export const getMainDomainOrgRedirect = (req: IncomingMessage | undefined) => {
```
If it needs just the URL, then it should just accept the URL. That is a clean code guideline.

Amit91848 replied:
Also wouldn't it be better to move it to a method inside `packages/lib/server/repository/organization.ts`

hariombalhara replied:
I think we shouldn't move it inside repository. Repository is supposed to be a DB layer. Application logic should remain separate IMHO.

But yeah I have thought about it multiple times that where should we hold all utility methods related to an entity. Maybe we could have lib/organization directory?

@zomars what do you think?

vijayraghav-io replied:
if we are using `organizationSettings.orgAutoAcceptEmail` then this `getMainDomainOrgRedirect.ts` is not required.
@hariombalhara please clarify does `orgAutoAcceptEmail` store only hostname always and not the email id, else validation needs to added to extract hostname from email id, what i could see in code is , it is inferred to hold hostname only, because of the naming i got this query.

hariombalhara replied:
Yeah, it is hostname only. The naming isn't clear. That Email in it actually means that all emails matching that domain will be automatically added to the organization.

vijayraghav-io replied:
ok sure thanks for clarifying

hariombalhara commented on apps/web/lib/getMainDomainOrgRedirect.ts:
```diff
@@ -0,0 +1,33 @@
+  const protocol = WEBAPP_URL.startsWith("https://") ? "https" : "http";
+  const originalUrl = new URL(req.url!, `${protocol}://${req.headers.host}`);
+
+  const domains = originalUrl.hostname.split(".");
```
Let's be clear with naming, hostname split doesn't give domains. May be we can call it, domainParts.

hariombalhara commented:
@vijayraghav-io We need this only for acme.cal.com and not acme.cal.com/anything_else. It wasn't clear from the ticket. I have updated it. Let's fix it. 

So, acme.cal.com and acme.cal.com/ -> Redirects to acme.com

Everything else remains same.

hariombalhara commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -175,6 +176,14 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
   }
 
+  const disableOrgSubdomainURL = team?.parent?.organizationSettings?.disableOrgSubdomainURL;
+  if (isValidOrgDomain && disableOrgSubdomainURL) {
+    const redirect = getMainDomainOrgRedirect(context.req);
+    if (redirect) {
+      return redirect;
+    }
+  }
```
We are going to have to identify the case where this route is for organization homepage(i.e. acme.cal.com) or a team page(e.g. acme.cal.com/marketing or cal.com/team1

vijayraghav-io committed:
updated to redirect as per orgAutoAcceptEmail and only in team slug

vijayraghav-io commented:
@hariombalhara , @Amit91848 updated
1. to redirect as per `orgAutoAcceptEmail`.
2. redirect only from org home page and team page.
3. updated without appending env var (ALLOWED_HOSTNAMES).
4. updated tests accordingly

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
nit/chore

hariombalhara commented on apps/web/lib/getMainDomainOrgRedirect.ts:
```diff
@@ -0,0 +1,45 @@
+
+const log = logger.getSubLogger({ prefix: ["lib", "getMainDomainOrgRedirect"] });
+export const getMainDomainOrgRedirect = (
+  req: IncomingMessage | undefined,
```
As I mentioned earlier, I don't think this method needs entire req. It just needs the URL

vijayraghav-io replied:
since we need host from req.headers.host , we need to pass req as param

hariombalhara commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -175,6 +178,28 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
+  if (!!organizationSettings) {
+    allowSEOIndexing = organizationSettings.allowSEOIndexing;
+
+    if (organizationSettings.disableOrgSubdomainURL) {
```
Because we are not actually disabling orgSubDomain URL. We are just redirecting the home page. I think we should rename it to something like 
`orgProfileRedirectsToVerifiedDomain`. We call i.cal.com page to be organization profile page that lists all teams of that organizations and other details.

So, it reads like this `organizationSettings.orgProfileRedirectsToVerifiedDomain`

hariombalhara replied:
I think we should update the copy(text) as well in the setting to better reflect it.

New text could be something like this.

'Disable public organization profile and redirect '
'Redirects deel.cal.com to deel.com'

cc @PeerRich. It is not a blocker, we can fix the text after the PR merge as well.

vijayraghav-io replied:
ok will rename var accordingly and update the description

vijayraghav-io commented on apps/web/lib/getMainDomainOrgRedirect.ts:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  const protocol = WEBAPP_URL.startsWith("https://") ? "https" : "http";
+  const originalUrl = new URL(req.url!, `${protocol}://${req.headers.host}`);
```
Will add a check for this (.url) too, thought since we are calling on page load -> getServerSideProps, if req is defined , then url should be present, checking for !req on first line

vijayraghav-io committed:
add check for req.url

hariombalhara commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -175,6 +178,28 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
+    if (organizationSettings.disableOrgSubdomainURL) {
+      const redirect = getMainDomainOrgRedirect(
+        context.req,
+        team?.children?.map((team) => team.slug || "").filter((slug) => !!slug),
```
Why do we need team.children? The requirement is to do the redirect only for organization profile.  It isn't applicable to Team Profile.


The motivation for the issue this PR is fixing is that people shouldn't be able to see a list of all the teams/users of the organization. That info is shown on organization profile page. So, the best was to redirect acme.cal.com to acme.com itself.

hariombalhara replied:
I think we could simplify the logic to just redirect to orgAutoAcceptEmail.

We don't need `getMainDomainOrgRedirect` function.

`redirectToVerifiedDomain({verifiedDomain})`

Function defined in this module.
```js
function getRedirectToVerifiedDomain({ verifiedDomain }: { verifiedDomain: string }) {
  return {
    redirect: {
      permanent: true,
      destination: `https://${verifiedDomain}`,
    },
  };
}
```

We know that this route is used in 3 cases. We just need to ensure that the function is called only in the 3rd case.
1. Team Profile Page
2. SubTeam Profile
3. Organization Profile Page

vijayraghav-io replied:
the same page i.e. team/[slug] is used for both org home page as well as for org/[team_slug] like (org/marketing).
here team in team?.children in org context is nothing but org.children (or teams under org).

so team?.children was used to validate team slug `acme.cal.com/[marketing]`

So do you mean, we need to redirect only when home page is visited `acme.cal.com` and no need for 
`acme.cal.com/marketing`

hariombalhara replied:
Yeah absolutely. See the updated description of the issue
![image](https://github.com/user-attachments/assets/98ebfa21-5aa2-4a0f-b166-6042b438d116)

vijayraghav-io replied:
got it

hariombalhara commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -31,10 +32,12 @@ const getTheLastArrayElement = (value: ReadonlyArray<string> | string | undefine
-    context.req,
-    context.params?.orgSlug ?? context.query?.orgSlug
-  );
+  const redirectDomainSlug = context.query?.redirectDomainSlug;
```
I don't think we need this parameter with the new approach. The only reason I think it could be useful is to just inform acme.com about the source of traffic which I believe can be determined from referrer.

vijayraghav-io replied:
we still need this as i mentioned earlier, if isValidOrgDomain = false  currentOrgDomain=null, this results to notfound page after redirect

hariombalhara replied:
Yeah that is true

vijayraghav-io committed:
updated orgProfileRedirectsToVerifiedDomain as per latest req

vijayraghav-io commented:
@hariombalhara , updated as per latest reqs
here is video for updated redirect functionality
https://www.loom.com/share/415c9dab0e54482594e3e7a9994d9772?sid=a313a1cf-db70-4875-8400-6d9457a896b5

Amit91848 commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -175,6 +185,25 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
     } as const;
   }
 
+  const organizationSettings = isOrgContext
+    ? team?.isOrganization
+      ? team?.organizationSettings
+      : team?.parent?.organizationSettings
+    : null;
+  let allowSEOIndexing = false;
+
+  if (!!organizationSettings) {
+    allowSEOIndexing = organizationSettings.allowSEOIndexing;
+
+    if (
+      team?.isOrganization &&
+      organizationSettings.orgProfileRedirectsToVerifiedDomain &&
+      organizationSettings.orgAutoAcceptEmail
+    ) {
+      return getRedirectToVerifiedDomain({ verifiedDomain: organizationSettings.orgAutoAcceptEmail });
+    }
+  }
+
```
Can this be moved before, to say line 93? We should prefer early returns, no point doing extra stuff if we are going to be redirecting

vijayraghav-io replied:
there is no much execution above this since it branches only in case of `if (!team)`
Anyway organized to have all redirects / returns at the begining at one place.

vijayraghav-io replied:
there is no much execution above this and branches out and executes more lines of code only in case of `!team` or no team found, anyway moved above and organized to have all redirects at the beginning

Amit91848 reviewed (changes requested):
These changes would also need tests

vijayraghav-io committed:
nit/chore: refactor for redirects

vijayraghav-io commented:
> These changes would also need tests

Since now we are having fn `getRedirectToVerifiedDomain()` inside getServerSideProps.tsx only tests are not added and it does nothing much than returning redirect{.. }.

Also currently there are no tests for entire getServerSideProps.tsx itself, we need to mock req and all, may be can be handled in a different PR

vijayraghav-io commented:
@hariombalhara / @Amit91848 , kindly let me know if anything is pending or any comments needs to addressed for this PR.

vijayraghav-io commented:
Reminding to please approve or share comments if any.

Amit91848 commented:
Can you fix the conflicts please @vijayraghav-io ?

vijayraghav-io commented:
> Can you fix the conflicts please @vijayraghav-io ?

sure will do it

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io commented:
will update for failing type checks after latest merge, in few mins

vijayraghav-io committed:
update for type checks after merge

vijayraghav-io commented:
@Amit91848 , fixed the conflicts, unit tests and type checks are also passing.
Thank you! 🙏

Amit91848 commented on packages/features/ee/organizations/pages/settings/appearance.tsx:
```diff
@@ -147,6 +153,32 @@ const OrgAppearanceView = ({
+      <SettingsToggle
+        toggleSwitchAtTheEnd={true}
+        title={t("disable_org_url_label")}
+        description={t("disable_org_url_description")}
```
AFAIK organizations cannot see what their autoAcceptEmail are. Can you pass the current org url and orgAutoAcceptEmail so that they exactly where they are being redirected to.

So description text will be dynamically generated different for different orgs like
`Redirects acme.cal.com to acme.com` for acme,  and then `Redirects dunder-mifflin.cal.com to dunder-mifflin.ai` etc.

Amit91848 replied:
I meant whatever `${orgSlug}.cal.com` and not current page url.

vijayraghav-io replied:
even if we say try to hard code to show current org url dynamically from autoAcceptEmail - getting `acme.com` and inserting `cal` in the middle, this may not be 100% accurate i feel, in all cases.

vijayraghav-io replied:
> I meant whatever `${orgSlug}.cal.com` and not current page url.

yeah ok, this must be possible will try to implement and update.

vijayraghav-io replied:
@Amit91848 , updated to show description dynamically and also tested once.

Amit91848 commented on packages/lib/server/repository/profile.ts:
```diff
@@ -36,6 +36,23 @@ const membershipSelect = Prisma.validator<Prisma.MembershipSelect>()({
+  allowSEOIndexing: true,
+  orgProfileRedirectsToVerifiedDomain: true,
+});
+const organizationWithSettingsSelect = {
+  id: true,
+  slug: true,
+  name: true,
+  metadata: true,
+  logoUrl: true,
+  calVideoLogo: true,
+  bannerUrl: true,
+  isPlatform: true,
+  organizationSettings: {
+    select: organizationSettingsSelect,
+  },
+};
```
NIT: Since it is the same
```suggestion
const organizationWithSettingsSelect = {
  ...organizationSelect,
  organizationSettings: {
    select: organizationSettingsSelect,
  },
};
```

Amit91848 reviewed (commented):
Changes look good, tested and it works as expected. Left some minor non blocking comments @vijayraghav-io

vijayraghav-io committed:
Update packages/lib/server/repository/profile.ts

Co-authored-by: Amit Sharma <74371312+Amit91848@users.noreply.github.com>

vijayraghav-io committed:
updated to show Redirect description dynamically

vijayraghav-io committed:
Merge branch 'feat_#16028_updateOrgProfile' of https://github.com/vijayraghav-io/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
declare organizationSelect before organizationWithSettingsSelect

Amit91848 commented on apps/web/lib/team/[slug]/getServerSideProps.tsx:
```diff
@@ -81,6 +88,26 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
+      organizationSettings.orgProfileRedirectsToVerifiedDomain &&
+      organizationSettings.orgAutoAcceptEmail
+    ) {
+      return getRedirectToVerifiedDomain({ verifiedDomain: organizationSettings.orgAutoAcceptEmail });
```
```suggestion
      return {
        redirect: {
          permanent: false,
          destination: `https://${verifiedDomain}`,
        },
      };
```

Function just returns the same thing and doesn't do anything additional, we can just keep it simple and return here itself.

vijayraghav-io replied:
yeah correct, i had just kept as suggested by hariombalhara, updated

vijayraghav-io committed:
removed function to redirect

Amit91848 reviewed (dismissed):
LGTM! Great work @vijayraghav-io 🔥🔥.

Will wait for another review before merging

vijayraghav-io commented:
> LGTM! Great work @vijayraghav-io 🔥🔥.
> 
> Will wait for another review before merging

Thankyou! @Amit91848 for your review, comments and approval 🙏 .
Understanding that the team is busy, Hoping another review/approval also happens soon as already other reviewers had provided their comments and i have also addressed them 🙏 .

hariombalhara committed:
Clean code in getSererSideProps

hariombalhara commented:
@Amit91848 Did this change in the repository to acheive what you were suggesting earlier
https://github.com/vijayraghav-io/cal.com/pull/1/files#diff-bb450a489482c4ed4d23859b7c335809e0eebd9088817bfa5e861f08e4f82b1cR201

hariombalhara reviewed (dismissed):
team/[slug]/[type] page doesn't have seoIndexing handled. There is no impact of toggling the seo setting on that page.

vijayraghav-io commented on apps/web/modules/users/views/users-public-view.tsx:
```diff
@@ -58,6 +58,13 @@ export function UserPage(props: InferGetServerSidePropsType<typeof getServerSide
   const isEventListEmpty = eventTypes.length === 0;
   const isOrg = !!user?.profile?.organization;
+
+  const allowSEOIndexing = isOrg
+    ? isOrgSEOIndexable
+      ? profile.allowSEOIndexing
+      : false
+    : profile.allowSEOIndexing;
```
The intention was to give priority to org (to which user belongs to) settings over user settings when SEOIndexing is disabled in Org.
And considering the security/privacy concern that if Org want to disable SEOIndexing, the user page will be disabled (regardless of user setting).

If only Org setting has to be enforced regardless of setting, then can be updated accordingly.

vijayraghav-io replied:
Currently,
_What if the User has seoIndexing disabled but Organization has it enabled. Do we want to enforce Organization setting there?_ - Since Org setting is enabled, no concern from org for enabling SEO Indexing, so User setting is taken 

_What if seoIndexing is disabled in Organization but enabled in user. Do we still want to enforce the org setting to user?_ - Here since it is disabling, Org Setting is prioritized and taken regardless of user setting.

Let me know, if need to update.

hariombalhara replied:
Yeah, we are discussing internally. I'll update here

vijayraghav-io replied:
may i know if there is any update.

vijayraghav-io replied:
@hariombalhara , Thanks for sharing,
What i infer is - UI should reflect the changes we are doing internally.
1. OrgSeo is disabled, User Seo is also set to false and switch (UI) is disabled.
2. OrgSeo is enabled, User Seo switch is enabled.

vijayraghav-io replied:
@hariombalhara , implemented this in UI .
if Org.AllowSeoIndexing is false, then user's AllowSeoIndexing Switch is disabled and shown false.
if Org.AllowSeoIndexing is true, then user's AllowSeoIndexing Switch is enabled and user's choice is shown.

vijayraghav-io replied:
@hariombalhara , to add - `Org.AllowSeoIndexing` is false by default.

hariombalhara replied:
By having it false by default, it effectively disables seo indexing for all by orgs by default. If some users of some organization have seo indexing enabled, it would disable it for them.

This is why we discussed about "No explicit value" which is when org.allowSeoIndexing is null


But I am thinking that orgs would normally not want to have SEO indexing enabled and it might be better to have it disabled automatically for all org members, unless organization allows it first.

vijayraghav-io committed:
Merge pull request #1 from calcom/improve-updateOrgProfile

Clean code in getSererSideProps

vijayraghav-io commented:
> @Amit91848 Did this change in the repository to acheive what you were suggesting earlier https://github.com/vijayraghav-io/cal.com/pull/1/files#diff-bb450a489482c4ed4d23859b7c335809e0eebd9088817bfa5e861f08e4f82b1cR201

Just to update, i had updated earlier to make early return as per Amit suggestion.
I guess the new changes is more of refactoring, have merged this as well.

vijayraghav-io commented:
> team/[slug]/[type] page doesn't have seoIndexing handled. There is no impact of toggling the seo setting on that page.

will update this, seoIndexing is disabled by default.

hariombalhara commented:
> > @Amit91848 Did this change in the repository to acheive what you were suggesting earlier https://github.com/vijayraghav-io/cal.com/pull/1/files#diff-bb450a489482c4ed4d23859b7c335809e0eebd9088817bfa5e861f08e4f82b1cR201
> 
> Just to update, i had updated earlier to make early return as per Amit suggestion.
> I guess the new changes is more of refactoring, have merged this as well.
> Thankyou! 🙏 

Yeah that's just refactoring.

vijayraghav-io committed:
update seoIndexing for team eventtype slug

vijayraghav-io commented:
@hariombalhara , Thanks for review 🙏 
for user-type-public-view.tsx, seoIndexing was already there before, only props needed update as per setting.
But for team/[slug]/[type], seoIndexing prop was not used, may be i guess that's how i had missed this, 
updated for team/[slug]/[type],

PFB video for same.
https://www.loom.com/share/fc62f79ad6df4ad68d3c1e646e08ec43?sid=aca04d60-e3b5-4a69-87ab-7d25c0b2f6cc

vijayraghav-io committed:
nit: remove redundant cond

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
updated to show org precedence in UI

vijayraghav-io commented:
@hariombalhara , please take this PR forward as all comments are addressed and implemented.

vijayraghav-io commented:
> @hariombalhara , 🙏  please take this PR forward as all comments are addressed and implemented.

vijayraghav-io commented:
@Amit91848 ,can you please take this PR forward ,all comments are addressed

alishaz-polymath committed:
Merge branch 'main' into feat_#16028_updateOrgProfile

hariombalhara commented:
@vijayraghav-io I am really sorry I haven't been able to review this PR yet. There are some really important things that are in higher priority going on.

@anikdhabal Would you like to take it over for review as it has been pending for a long long time on me

vijayraghav-io commented:
> @vijayraghav-io I am really sorry I haven't been able to review this PR yet. There are some really important things that are in higher priority going on.
> 
> @anikdhabal Would you like to take it over for review as it has been pending for a long long time on me

@hariombalhara , Thankyou! for commenting 🙏 , I understand.
Hoping if this can takeoff soon.

vijayraghav-io commented:
> @vijayraghav-io could you pls fix the conflicts

Sure, will do.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io commented:
@anikdhabal , resolved merge conflicts

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io committed:
added e2e tests

vijayraghav-io committed:
nit

keithwillcode commented:
@vijayraghav-io You mind fixing the conflicts please? We are looking to merge this one soon for you! Thank you 🙏 

@hariombalhara Do you mind giving this a once over again now that changes have been made?

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16028_updateOrgProfile

vijayraghav-io commented:
@keithwillcode , @hariombalhara  resolved conflicts.

hariombalhara commented on apps/web/server/lib/[user]/[type]/getServerSideProps.ts:
```diff
@@ -219,6 +219,12 @@ async function getUserPageProps(context: GetServerSidePropsContext) {
     } as const;
   }
 
+  const allowSEOIndexing = org
```
Followup: Abstract it out into something like `packages/lib/isSeoIndexableForUser` and reuse it at all places where we derive user's seoIndexing setting from org and user.

hariombalhara reviewed (approved):
Thanks for your patience here @vijayraghav-io !! Great work !!

let's merge it.

I have one followup comment, that we can address later.

vijayraghav-io commented:
> Thanks for your patience here @vijayraghav-io !! Great work !!
> 
> let's merge it.
> 
> I have one followup comment, that we can address later.

Thank you!  @hariombalhara 🙏 , 
sure will address the followup comments.
